Add Consume bool channel to consumer, created State machine and Start…

DavidSubiros opened the pull request:
…/Stop consuming accordingly

### What

Describe what you have changed and why.

### How to review

Describe the steps required to test the changes.

### Who can review

Describe who worked on the changes, so that other people can review.

ChrysmOre commented on consumer_state.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+	if !transitionAllowed {
+		return fmt.Errorf("state transition from %s to %s is not allowed", sm.state.String(), newState.String())
```
When you have a type that implements the `stringer` interface you don't need to explicitly call the String() function when using Printf etc :- you can just pass `sm.state` and `newState` directly here

ChrysmOre commented on error.go:
```diff
@@ -0,0 +1,84 @@
+// LogData implements the DataLogger interface which allows you extract
+// embedded log.Data from an error
+func (e *Error) LogData() map[string]interface{} {
+	if e.logData == nil {
```
I always designed the unwrapLogData function to handle `nil` logData to prevent having empty structs in the logs

ChrysmOre commented on error.go:
```diff
@@ -0,0 +1,84 @@
+	for _, d := range data {
+		for k, v := range d {
+			if val, ok := logData[k]; ok {
+				if val != v {
```
I noticed somewhere you'd added a `!reflect.DeepEqual` instead of `!=` here, I thought that was a good idea.

DavidSubiros commented on config_security.go:
```diff
@@ -0,0 +1,92 @@
+		if cert, err = tls.X509KeyPair(
+			[]byte(expandNewlines(tlsConfig.ClientCert)),
+			[]byte(expandNewlines(tlsConfig.ClientKey)),
+		); err != nil {
```
err is the named var, but I will change it to make it clearer.

DavidSubiros commented:
> There are some instances of named return values and errors without context I pointed out, there are more instances and I didn't comment on every single one but the comments still apply there too.

Pushed a new commit where I added context to all errors, and fixed the tests accordingly. I also fixed/improved a couple more things:
- Tried to reduce logging and make it consistent
- Added handler/batch-handler go-routines to Closer wait group, so that Close() waits until the handler go-routines finish (not only the sarama/consume/stopped/initiliased loops)
